### ISSUE TITLE
wrong implementation/definition of ftol_iter

### ISSUE THREAD

**Issue by D4T4-80Y** [ljvmiranda921/pyswarms#447]
**Description**

When **ftol** is used, for example 1e-10, pyswarm stops when `swarm.best_cost` is lower than `ftol`  for `ftol_iter` times. However, the correct naming of `ftol_iter` should be `stall_iter`.  This means that the search stops when best_cost isn't improved after x iterations, or when it is below `ftol`. However, when I search for a solution, I need to stop the search when the best_cost isn't improved after `x` iterations or when it is below `ftol` (must happen only once).

The current naming and implementation is confusing and doesn't make sense. That's why I thought I had to set `ftol_iter=1` which caused stopping after the first iteration and having premature solution ([0.03840509 0.00329153] instead of [-9 -9]).

In this regard, the  default value of ftol_iter should be big enough (at least 100) to avoid having premature solution. ftol_iter=1 is unusual and very small. 

This is what I got with ftol=1e-10 and ftol_iter=100. As can be seen, it must stop after iteration 199.

```
196 = {float64} 1.543800321250907e-10
197 = {float64} 1.543800321250907e-10
198 = {float64} 1.543800321250907e-10
199 = {float64} 1.2788497031507338e-12   <-- stop here
200 = {float64} 1.2788497031507338e-12
201 = {float64} 1.2788497031507338e-12
202 = {float64} 1.2788497031507338e-12
203 = {float64} 1.2788497031507338e-12
204 = {float64} 1.2788497031507338e-12
```

**Sample code**
A sample code is attached. The function to be optimized is a modified sphere function centered at [-9,-9]. The real cost is **zero**.

[ftol_test.zip](https://github.com/ljvmiranda921/pyswarms/files/5940520/ftol_test.zip)

**Expected behavior**
As I said, I expect that the optimizer stops after `ftol_iter` iterations when there is no improvement (best_cost stalls) or it stops **instantly** when the best_cost is below `ftol`. With the current implementation, optimizer only stops after iteration around 300, after stalling for 100 iterations but with ftol around 2e-16, which is much lower than given ftol. However, it must stop somewhere around iterations 200.

**Environment (please complete the following information):**
 - PySwarms Version 1.3.0

Thank you.
DATA-BOY

**Comment by D4T4-80Y** [comment]
Another reason for saying that `ftol_iter` is not a good naming and it should be renamed to `stall_iter` is if a new option, xtol, is implemented, then a new name makes sense. It is absolutely pointless to define a new option, `xtol_iter`.

`xtol` defines a minimum accepted threshold for changes in positions, X, so the search can be stopped if the changes in postions below the given value.

**Comment by ljvmiranda921** [comment]
Hi @D4T4-80Y , thanks for bringing this up. So far only one person commented on the naming of `ftol_iter`, and I don't want to run the risk of creating a lot of breaking changes by changing the parameter. I suggest that it may be better to just update the documentation or docstring of `ftol_iter` rather than changing it entirely. Would you mind opening up a PR for that?

**Comment by D4T4-80Y** [comment]
Why don't you deprecate ftol_iter in the next versions and create a new and more intuitive parameter?

**Comment by msat59** [comment]
Hi @D4T4-80Y and @ljvmiranda921.
This is almost what I complained about few month ago. @D4T4-80Y  raised and explained it in a better way.

As I have had implemented this fix on my machine, I am ready to fix it and even implement new feature, xtol, and even `MaxStallTime` and `MaxTime` like MATLAB particleswarm().

**Comment by ljvmiranda921** [comment]
> As I have had implemented this fix on my machine, I am ready to fix it and even implement new feature, xtol, and even MaxStallTime and MaxTime like MATLAB particleswarm().

Hi @msat59 , I will appreciate the PR, thanks a lot. Let's do the following instead:
* Rename `ftol_iter` to `stall_iter` and update the documentation (this should be a quick fix)
* Implement `xtol` with good defaults

Let's target this for a March release, sounds good?
Thanks for the issue @D4T4-80Y and would be great for you to help out atleast on action item 1

**Comment by msat59** [comment]
> > As I have had implemented this fix on my machine, I am ready to fix it and even implement new feature, xtol, and even MaxStallTime and MaxTime like MATLAB particleswarm().
> 
> Hi @msat59 , I will appreciate the PR, thanks a lot. Let's do the following instead:
> 
> * Rename `ftol_iter` to `stall_iter` and update the documentation (this should be a quick fix)
> * Implement `xtol` with good defaults
> 
> Let's target this for a March release, sounds good?
> Thanks for the issue @D4T4-80Y and would be great for you to help out atleast on action item 1

Hi @ljvmiranda921 .

Thanks for your confirmation about changing the name. I personally use intuitive parameters and variables names in my codes. So I prefer `max_stall_iter`.  I believe `stall_iter` doesn't mean anything at first glance. In the same fashion, I really looking forward to renaming `iters` in `optimize` method to `max_iter`, which is more intuitive.

I know it's a big change, but we can keep all old params, and send warning to users that x,y and z parameters will be deprecated in version zzz.

I listed my suggestions so you confirm what changes are allowed:

1- renaming `ftol_iter` to `max_stall_iter`
2- renaming `iters` in optimize method to `max_iter`
3- setting a default value for `max_iter`. According to MATLAB documentation, `200*dimensions` is a good value. ( _I'd checked it in few cases and it had worked well_ ).
As a result, a novice user can start optimizing quickly, without thinking about max_iter, just like this:
`cost, pos = optimizer.optimize(f_obj)
4- implementing `xtol` (_ you approved previously_ )

There are few more new features that I can implement such as max_stall_time, max_time, returning the exit_flag/exit_reason, and updating the LOGGER to show the termination reason.

I will discuss them in a separate topic.

**Comment by D4T4-80Y** [comment]
> 
> 
> > As I have had implemented this fix on my machine, I am ready to fix it and even implement new feature, xtol, and even MaxStallTime and MaxTime like MATLAB particleswarm().
> 
> Hi @msat59 , I will appreciate the PR, thanks a lot. Let's do the following instead:
> 
>     * Rename `ftol_iter` to `stall_iter` and update the documentation (this should be a quick fix)
> 
>     * Implement `xtol` with good defaults
> 
> 
> Let's target this for a March release, sounds good?
> Thanks for the issue @D4T4-80Y and would be great for you to help out atleast on action item 1

Hi @ljvmiranda921.

I'm not sure that I can assist properly in that case as @msat59 has had already a working code. However, I am more than happy to assist if needed.

**Comment by nishnash54** [comment]
@D4T4-80Y You make a great point about the having a `stall` option. I don't think this is currently a feature present in pyswarms. It'll be a great addition. Before we make any decisions about a new feature (which I would love to implement) I just want to make sure we are on the same page.

**`ftol` and `ftol_iter`**
The initial idea of the `ftol` parameter is to stop once the set tolerance is reached. In quite a few use cases this is enough. As swarm algorithms are many a times quite volatile at the start (especially with random initialization) we can get cases where we reach the required tolerance without being sure that this value can be sustained (enough convergence has occurred). To solve this issue, the `ftol_iter` parameter was introduced. The concept is taken from deep learning where we require the model to achieve the set threshold over the next few iteration to stop further training.

You can check the implementation [here](https://github.com/ljvmiranda921/pyswarms/blob/3277a3d7e849972a3195687ee89e6d83bd99592f/pyswarms/discrete/binary.py#L220)

**`stall_iter`**
From your example and explanations, what you are looking for is that the algorithm stops it you encounter a stall situation (which is **no improvements** in tolerance over `stall_iter` iteration.) So in this case the `stall_iter` will be a new parameter and this will work irrespective of any user defined tolerance values.

```py

# params ftol=1e-12, ftol_iter=5
196 = {float64} 1.543800321250907e-10
197 = {float64} 1.2788497031507338e-12   # <-- stop here (if only ftol)
198 = {float64} 1.543800321250907e-10
199 = {float64} 1.2788497031507338e-12
200 = {float64} 1.2788497031507338e-12
201 = {float64} 1.2788497031507338e-12
202 = {float64} 1.2788497031507338e-12
203 = {float64} 1.2788497031507338e-12   # <-- stop here (with folt and ftol_iter)
204 = {float64} 1.2788497031507338e-12
```

@ljvmiranda921 @msat59 Are we on the same page here? We should update the parameters and docs to be more informative to the user.

```yaml
ftol_iter : int
    [-] number of iterations over which the relative error in objective_func(best_pos) ...
    [+] Makes sure the defined threshold is achieved for `ftol_iter` consecutive iterations.
    Default is :code:`1`
```

**Comment by D4T4-80Y** [comment]
Hi @nishnash54.

The **stall** means **no improvement** in the best_cost no matter what its value is, which means it stays constant in two consecutive iterations . In other word, diff=0 which is always less than `ftol`.

> The concept is taken from deep learning where we require the model to achieve the set threshold over the next few iteration to stop further training.

What you are referring to is called **Early Stopping** (mostly used in Neural Networks) and the main purpose is to avoid overfitting while training a model. So the algorithm stops training once the model performance stops improving on the validation dataset for the given number of iterations. **THIS IS HOW ftol_iter WORKS AT THE MOMENT.**

**Comment by msat59** [comment]
Hi @nishnash54  and @D4T4-80Y.

Thanks for your explanations. My concern is the practicality of `ftol_iter ` in the optimization context. I think we can have both stall_iter and ftol_iter, but we must be clear about its purpose. Let's define the purpose of PySwarm first, and then talk about the features. We can then conclude better.

1- PySwarms is an **_optimization_** package
or
2- PySwarms is a **_machine learning_** package

To my understanding, Pyswarms is an optimization package, which uses the Particle Swarm Optimization algorithm, so we don't need to have deep learning or other machine learning packages features. We are not building a regression/classification model here. We are trying to minimize a cost function as fast as possible. If someone wants to use PySwamrs in a machine learning context, one should write some codes for his/her needs. :wink:

In this regard, the definition and implementation of `ftol_iter` and  `stall_iter` are different in different use cases:

1- **Optimization**: stop the search once the difference is less than `ftol` **excluding** no improvements (so no need to ftol_iter, or can say ftol_iter=1), **or** consecutive iterations with no improvement, `stall_iter`, is reached. 

2- **Machine learning (early stopping)**: stop the search when the difference is less than `ftol` **including** no improvements, for `ftol_iter` of times consecutively. 

That's the source of all issues that @D4T4-80Y and I have. The machine learning approach (current implementation) is not compatible with numerical optimization approach; it does not consider no improvements (stalls) separately so it results in an immature solution with the default value of `ftol_iter=1`, and increasing the ftol_iter value may result in processing all iterations, which makes the use of `ftol` pointless and may be time-consuming to process all iterations.

In PSO algorithm, as the search proceeds and the algorithm performs more iterations, the best_cost may stall so there may not be any changes in it for several iterations. Then one or few improvements may occur and then the best_cost stalls for several iterations again. In real-life problems, we don't know when a new improvement happens, how many times it will be below ftol, and for how many iterations the best_cost stalls. So why should one bother to introduce more complexity using ftol_iter>1 and miss a possible solution once ftol was met?

In all numerical packages that I have ever used since 2001 (not in machine learning algorithms), I haven't seen any early stopping feature; none of them has implemented or looking for a stop criterion like what we have now; because they are not training a model so they don't care about overfitting. Once the difference is below the acceptable threshold, they are done.
 
The story is different in the context of machine learning. One may use PySwarms to tune a model hyperparameters. In this case, one may be interested in having an early stopping feature, without trying to implement it by him/herself (which is not a big deal).

@nishnash54, I completely understand what you say and it may be a good idea to have both stall_iter and ftol_iter; but let's consider the purpose and application of PySwarms. I believe that it is pointless to have early stopping in an optimization package because both stall_iter and early stopping are conflicting and may not be possible to use both at the same time.

Please let me know what is the real application and purpose of ftol_iter? 
@nishnash54 
@ljvmiranda921 

## Is ftol_iter an early stopping feature? ##

## If yes, do we need early stopping feature in PySwarms? ##

## If No, why should use ftol_iter>1? ##


Based on the decision made, I can work on stall_iter. At the moment, if I correct the behavior of ftol_iter and implement stall_iter, it will not be **early stopping** any more, and practically, ftol_iter > 1 is pointless to use. So ftol_iter should be removed to avoid any confusions in using stall_iter or ftol_iter.

If the early stopping is needed, it's another story. The ftol_iter may need to be renamed to something meaningful, and one should be enabled while optimizing, ftol_iter or stall_iter.

Thanks, guys.

**Comment by ljvmiranda921** [comment]
Hi all , 

So to round it up, the problem is on how we can incorporate @D4T4-80Y 's required functionality into Pyswarms, because as it turns out his/her expectation of `ftol_iter` is different. We now have three options:

* Option A: implement `stall_iter` that is different from `ftol_iter`. While **keeping** the latter's implementation the same.
* Option B: implement `stall_iter` that is different from `ftol_iter`. Update the implementation of `ftol_iter` into something different, rename the current implementation of `ftol_iter`.
* Option C: require early stopping, which might be a different use-case and implementation altogether.

I believe we're leaning towards Option B? I'm ok renaming `ftol_iter` (cc: @nishnash54 )
If this is good then I think we can proceed

**Comment by msat59** [comment]
Thanks @ljvmiranda921 for the conclusion.

Option B is the best one to me, which satisfies @D4T4-80Y and @nishnash54, and both features won't have any conflict. We can also keep `ftol_iter` name as it is and update its definition as @nishnash54 posted.

**Comment by nishnash54** [comment]
@ljvmiranda921 @msat59 Let's go ahead with **option B**. Seems to solve the issues. Let me know how you want to rename `ftol_iter`. I'll update the documentation accordingly.

@msat59 If you have already implemented the `stall_iter` let's use that, otherwise I can work on a new PR.

**Comment by msat59** [comment]
> @msat59 If you have already implemented the `stall_iter` let's use that, otherwise I can work on a new PR.

@nishnash54, I have implemented it already. Just need to paste the code into BinaryPSO, and test it. I will submit it by this weekend.

**Comment by ljvmiranda921** [comment]
Thanks for everyone's cooperation! :bow:

**Comment by msat59** [comment]
Guys, there is an issue in creating pools. 
Technically, no pool should be created for `n_processes=1`; but current implementation creates **one** pool, which costs around 1 second without having any benefit. It's not much; but the code should be fixed.

```
n_processes = None if n_processes==1 else n_processes
```

Should I fix it separately or can I fix it in this PR?
@ljvmiranda921

**Comment by ljvmiranda921** [comment]
Hi @msat59 , let's do it in another PR instead :+1:

**Comment by nishnash54** [comment]
Instead of `stall_iter` maybe we could use `n_iter_no_change`. It's a bit lengthy, but this naming is also followed by **sci-kit learn** methods and is more informative for the user.

**Comment by msat59** [comment]
@D4T4-80Y
@ljvmiranda921 
@nishnash54 

Guys, with modifications in default values of `ftol` and `ftol_iter` and the description of `ftol_iter`, it becomes what I implemented in a different way as the `stall_iter`.  :grimacing:

I had replicated the Matlab Toolbox code that I had, so it had looked different. Also, the description of ftol_iter had been confused me.  Today, I was doing the final checks so I decided to fix the numpy seed to get replicable results ([Issue #235](https://github.com/ljvmiranda921/pyswarms/issues/235)). Then I noticed that I can convert my code to something like ftol_iter implementation; `if all(stall_history): ...` and then... :boom:

Damn, the fact is that setting `stall_iter=50` and` ftol_iter=1` exactly gives the same results as `ftol_iter=50`.

So if we change the default value of `ftol_iter=50` and `ftol=1e-10', change their definitions to something like below description, and modify the code slightly (makes it more readable), everything is resolved.

```
ftol : float
   relative error in objective_func(best_pos) acceptable for convergence. To disable it in order to process all iterations,
   use ftol=None. Default is :code:`1e-6`

ftol_iter : int
    number of consecutive iterations over which the relative change in objective_func(best_pos) is stopped or less than ftol. 
    Default is :code:`50`
```

and the modifications are:

at the beginning of the **optimize** method:
```
ftol = -np.inf if ftol is None
```

```
ftol_history.append(delta)
    if i > self.ftol_iter:
        if all(ftol_history):
            break
```

So I will create a PR with these changes if it is fine with you guys.

one more thing:

> Instead of `stall_iter` maybe we could use `n_iter_no_change`. It's a bit lengthy, but this naming is also followed by **sci-kit learn** methods and is more informative for the user.

If we want the parameters to be more informative, we should rename `Iters` to `max_iter` and `ftol_iter` to `stall_iter`!

**Comment by msat59** [comment]
Another reason for insisting on renaming `ftol_iter` to `stall_iter` or `MaxStallIter` is the implementation of `fmin`, which is the stop condition when `best_cost < fmin` and will use the same parameter. If we keep `ftol_iter`, it is not intuitive when using `fmin`. If we rename it to `stall_iter`, then both `ftol` and `fmin` can use the same parameter for checking the stop condition.
